File: java/SearchForARange.java
//this can be refactored. but i'm head to lunch. XD
public class Solution {
    public int[] searchRange(int[] A, int target) {
        int[] result = new int[2];
        int exist = checkExist(A, target);
        if (exist == -1) {
            result[0] = -1;
            result[1] = -1;
            return result;
        }
        result[0] = this.searchLeft(A, 0, exist);
        result[1] = this.searchRight(A, exist, A.length-1);
        return result;
    }
    
    public int searchLeft(int[] A, int left, int right) {
        int target = A[right];
        while(left <= right) {
            int mid = (left + right) / 2;
            if ((mid == 0 || A[mid-1] != target) && A[mid] == target) {
                return mid;
            }
            if (A[mid] != target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right;
    }
    
    public int searchRight(int[] A, int left, int right) {
        int target = A[left];
        int mid = (left + right) / 2;
        if ((mid == A.length - 1 || A[mid+1] != target) && A[mid] == target) {
            return mid;
        }
        if (A[mid] != target) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
        return left;
    }
    
    public int checkExist(int[] A, int target) {
        int left = 0;
        int right = A.length - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (A[mid] == target) {
                return mid;
            } else if (A[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return -1;
    }
}

File: archive/cpp/array/SearchForARange.cpp
class Solution {
public:
    vector<int> searchRange(int A[], int n, int target) {
        vector<int> result;
        int mid = this->checkExist(A, n, target);
        if (mid == -1) {
            result.push_back(-1);
            result.push_back(-1);
        } else {
            result.push_back(this->searchLeft(A, target, mid));
            result.push_back(this->searchRight(A, n, target, mid));
        }
        return result;
    }
    
public:
    int checkExist(int A[], int n, int target) {
        int low = 0;
        int high = n - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (A[mid] == target) {
                return mid;
            } else if (A[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
    
public:
    int searchLeft(int A[], int target, int mid) {
        int low = 0;
        int high = mid;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (A[mid] == target) {
                high = mid - 1;
            } else {
                if (A[mid+1] == target) {
                    return mid + 1;
                } else {
                    low = mid + 1;
                }
            }
        }
        return low;
    }

public:
    int searchRight(int A[], int n, int target, int mid) {
        int low = mid;
        int high = n - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (A[mid] == target) {
                low = mid + 1;
            } else {
                if (A[mid-1] == target) {
                    return mid - 1;
                } else {
                    high = mid - 1;
                }
            }
        }
        return high;
    }
};
